stdlib/src/memory/boxed.mojo

```diff
@@ -0,0 +1,143 @@
+# ===----------------------------------------------------------------------=== #
+# Copyright (c) 2024, Modular Inc. All rights reserved.
+#
+# Licensed under the Apache License v2.0 with LLVM Exceptions:
+# https://llvm.org/LICENSE.txt
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ===----------------------------------------------------------------------=== #
+from memory import UnsafePointer, stack_allocation, memcpy
+
+
+struct Boxed[T: AnyType, address_space: AddressSpace = AddressSpace.GENERIC]:
+    """A safe, owning, smart pointer.
+
+    This smart pointer is designed for cases where there is clear ownership
+    of the underlying data, and restricts access to it through the lifetime
+    system such that no more than one mutable alias for the underlying data
+    may exist. Consider Boxed[T] over UnsafePointer[T] where possible.
+
+    Parameters:
+        T: The type to be stored in the Boxed[].
+        address_space: The address space that the allocation behind this box resides in.
+    """
+
+    var _inner: UnsafePointer[T, address_space]
+
+    fn __init__[
+        T: Movable
+    ](inout self: Boxed[T, AddressSpace.GENERIC], owned value: T):
```

Comment:
@szbergeron why not relying on the `address_space` param, which is `AddressSpace.GENERIC` by default?

```suggestion
    ](inout self: Boxed[T, address_space], owned value: T):
```

Reply:
This is waiting on design decisions regarding custom allocators. It's unclear what could substitute for `malloc` as a sane default for allocating in any addrspace other than GENERIC. I did want to at least stub that out, but a long discussion in https://discord.com/channels/1087530497313357884/1285357816030298112 mostly arrived at the conclusion that

1. Small allocations with Boxed[] don't make that much sense on things like GPUs (no good way of doing a traditional stack)
2. Custom allocators themselves need a bunch of design work to find the right abstraction
3. It might not make sense to have such an open constructor, instead of treating it has relatively hidden such that "deeper" libraries can return already constructed ones that nicely follow some rules for more efficient allocation (I could see Boxed[] being a relatively generic "this is a handle on some data that can be released using an allocator api")